### PR TITLE
Optimize MLA generate_mask

### PR DESCRIPTION
**Pull request by zcjhao**
# Description

Optimize MLA generate_mask with TPU-native vectorized threshold cutoff and configurable exact-k pruning
  
  Replace the broadcast comparison in `generate_mask` with a vectorized threshold cutoff comparison, adding the
`indexer_mask_exact_topk` configuration flag to dynamically support raw thresholding or exact-k prefix-sum pruning.
  
  - Two options:
    - Option 1 (`indexer_mask_exact_topk = False`): Executes raw vectorized threshold comparison. Allows > k tokens under boundary ties.
    - Option 2 (`indexer_mask_exact_topk = True`, default): Runs a cumulative prefix-sum pruning scan (`jnp.cumsum`) over active
boundaries, keeping exactly the first k selected tokens and discarding tied overflows.
      
  ### Context & Implementation Details
  * **The Problem:** The baseline implementation `(jnp.arange(s) == topk_indices[..., None]).any(axis=-2)` executes an all-pairs broadcast index comparison. 
  * **The Solution:** We leverage the $k$-th score from `top_k` as a decision threshold (`cutoff = topk_values[..., -1:]`).
Value-thresholding (`score >= cutoff`) converts the random index-matching query into an element-to a single value comparison.
  * **Configurable Tie Handling:** Under boundary ties, raw thresholding unmasks all tied tokens (allows $>k$ tokens). To allow strict density control, Option 2 executes a sequence-wide `cumsum` over active threshold boundaries. This counts and filters the unmasked elements by selection rank.
      
  # Tests
  
  ### Correctness & Boundary Verification
  Added unit tests in `tests/unit/attention_test.py` asserting that vectorized threshold cutoff masking:
  * Matches exact top-k selection when scores are unique (`test_generate_mask_threshold_equivalence`).
  * Unmasks $k+x$ elements under ties when `indexer_mask_exact_topk = False` (raw cutoff,
`test_generate_mask_threshold_ties_raw`).
  * Unmasks exactly $k$ elements under ties when `indexer_mask_exact_topk = True` (prefix-sum pruning,
`test_generate_mask_threshold_ties_exact_k`).
  * Edge case when `k` is less than sequence length `test_generate_mask_sequence_smaller_than_k`.

* `test_generation_deepseek32_approx_indexer` in `tests/unit/deepseek32_vs_reference_test.py` passed

 ### Profiles
 * [Xprof Baseline](https://xprof.corp.google.com/overview_page/zjiahao-14979029504970122278): Step time 74 s
 * [xprof indexer_mask_exact_topk: true](http://xprof/?session_id=zjiahao-12189676620266372785): Step time 60 s
 * [xprof indexer_mask_exact_topk: false](http://xprof/?session_id=zjiahao-10104111445293292642): Step time 56 s

  # Checklist

Before submitting this PR, please make sure (put X in square brackets):

[✓] I have performed a self-review of my code.
[✓] I have necessary comments in my code, particularly in hard-to-understand areas.
[✓] I have run end-to-end tests tests and provided workload links above if applicable.
[✓] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: src/maxtext/layers/attention_mla.py [hunk 1]
```diff
@@ -221,26 +221,29 @@ def apply_partial_rope(
...
     val_true = jnp.array(0.0, dtype=self.dtype)
     val_false = jnp.array(DEFAULT_MASK_VALUE, dtype=self.dtype)
-    return jnp.where(is_topk, val_true, val_false)
+
+    if self.config.indexer_mask_exact_topk:
+      # Prune ties by keeping only the first k unmasked tokens along sequence dimension
+      selected_rank = jnp.cumsum(raw_mask.astype(jnp.int32), axis=-1)
+      return jnp.where((raw_mask) & (selected_rank <= topk_values.shape[-1]), val_true, val_false)
+    else:
+      # Raw threshold cutoff masking (optional: enables speedups, but may unmask > k tokens under indexer scores ties)
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply src/maxtext/layers/attention_mla.py]: Could you cross check and also keep the original implementation if possible?

File: src/maxtext/layers/attention_mla.py [hunk 2]
```diff
@@ -367,16 +370,16 @@ def __call__(
 
     # TopK selection based on index score
     if self.config.indexer_use_approx_top_k:
-      _, topk_indices = jax.lax.approx_max_k(
+      topk_values, topk_indices = jax.lax.approx_max_k(
```
Comment (RissyRan) [inline src/maxtext/layers/attention_mla.py:389]: Wondering if you see any perf regression due to this change, including throughput and memory usage?

File: tests/unit/attention_test.py [hunk 3]
```diff
@@ -2230,6 +2230,116 @@ def full_indexer_loss_fn(inputs_q, inputs_kv, low_rank_q, mla, sparse_training=s
...
+        dtype=dtype,
+    )
+
+    topk_values, _ = jax.lax.top_k(scores, k=k)
+    mask = mla.indexer.generate_mask(scores, topk_values)
+
+    self.assertFalse(jnp.isnan(mask).any())
+    self.assertEqual(jnp.sum(mask[0, 0] == 0.0), 3)  # Exactly 3 tokens (exact k) unmasked
+    self.assertEqual(jnp.sum(mask[0, 1] == 0.0), 3)  # Exactly 3 tokens (exact k) unmasked
+
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply tests/unit/attention_test.py:2354]: Please consider the comment above.

[the remark this answers is not in the document]
Comment (zcjhao) [reply tests/unit/attention_test.py:2354]: yes!

File: src/maxtext/layers/attention_mla.py (hunk 2 above)
Comment (zcjhao) [reply src/maxtext/layers/attention_mla.py:389]: In the profiling I see reduced step time, with 1Gib peak memory increase. No perf regression observed.

File: src/maxtext/layers/attention_mla.py [hunk 4]
```diff
@@ -221,26 +221,45 @@ def apply_partial_rope(
...
 
     Args:
-        topk_indices: [b, t, k] int - The indices to keep.
-        s: int - The total size to select from.
+        indexer_score: [b, t, s] float - The computed relevance scores for all tokens.
+        topk_values: [b, t, k] float - The top-k selected score values from jax.lax.top_k.
 
     Returns:
-        mask: [b, t, s] - `0.0` at topk_indices, `DEFAULT_MASK_VALUE` (large negative) elsewhere.
+        mask: [b, t, s] - `0.0` for top-k selected elements, `DEFAULT_MASK_VALUE` elsewhere.
```
Comment (gobbleturk) [inline src/maxtext/layers/attention_mla.py:232]: consider lazy mask as followup

**Review by gobbleturk — APPROVED** [review]
Awesome tests and edge case consideration!

**Review by RissyRan — APPROVED** [review]
